config/initializers/http_client_proxy.rb

```diff
@@ -0,0 +1,24 @@
+Rails.application.configure do
+  config.x.http_client_proxy = {}
+  if ENV['http_proxy'].present?
+    proxy = URI.parse(ENV['http_proxy'])
+    raise "Unsupported proxy type: #{proxy.scheme}" unless ["http", "https"].include? proxy.scheme
+    raise "No proxy host" unless proxy.host
+
+    host = proxy.host
+    host = host[1...-1] if host[0] == '[' #for IPv6 address
+    config.x.http_client_proxy[:proxy] = { proxy_address: host, proxy_port: proxy.port, proxy_username: proxy.user, proxy_password: proxy.password }.compact
+  end
+
+  config.x.access_to_hidden_service = ENV['ALLOW_ACCESS_TO_HIDDEN_SERVICE'] == 'true'
+  config.x.hidden_service_via_transparent_proxy = ENV['HIDDEN_SERVICE_VIA_TRANSPARENT_PROXY'] == 'true'
+end
+
+module Goldfinger
+  def self.finger(uri, opts = {})
+    to_hidden = /\.(onion|i2p)(:\d+)?$/.match(uri)
```

Comment:
This isn't specific either to I2P or Tor: someone might want to federate with servers on Freenet, or on some mesh network or private network that requires a proxy for access.

It seems that there are two concerns: list of TLDs for which TLS is not enforced, and list of TLDs which are accessed via a proxy.